python/tools/recorded-call-graph-metrics/ql/lib/RecordedCalls.qll

```diff
@@ -0,0 +1,269 @@
+      // AST call (such as for list-comprehensions).
+      //
+      // So allowing partial matches is not safe, since we might end up matching a
+      // recorded call not in the AST together with an unrecorded call visible in the
+      // AST.
+    )
+  }
+}
+
+/** The XML data for the callee part a recorded call. */
+abstract class XMLCallee extends XMLElement { }
+
+/** The XML data for the callee part a recorded call, when the callee is a Python function. */
+class XMLPythonCallee extends XMLCallee {
+  XMLPythonCallee() { this.hasName("PythonCallee") }
+
+  string get_filename_data() { result = this.getAChild("filename").getTextValue() }
+
+  int get_linenum_data() { result = this.getAChild("linenum").getTextValue().toInt() }
+
+  string get_funcname_data() { result = this.getAChild("funcname").getTextValue() }
+
+  Function getACallee() {
+    result.getLocation().hasLocationInfo(this.get_filename_data(), this.get_linenum_data(), _, _, _)
+    or
+    // if function has decorator, the call will be recorded going to the first
+    result
+        .getADecorator()
+        .getLocation()
+        .hasLocationInfo(this.get_filename_data(), this.get_linenum_data(), _, _, _)
+  }
+}
+
+/** The XML data for the callee part a recorded call, when the callee is a C function or builtin. */
+class XMLExternalCallee extends XMLCallee {
+  XMLExternalCallee() { this.hasName("ExternalCallee") }
+
+  string get_module_data() { result = this.getAChild("module").getTextValue() }
+
+  string get_qualname_data() { result = this.getAChild("qualname").getTextValue() }
+
+  Builtin getACallee() {
+    exists(Builtin mod |
+      mod.isModule() and
+      mod.getName() = this.get_module_data()
+    |
+      result = traverse_qualname(mod, this.get_qualname_data())
+    )
+  }
+}
+
+/**
+ * Helper predicate. If parent = `builtins` and qualname = `list.append`, it will
+ * return the result of `builtins.list.append`.class
+ */
+private Builtin traverse_qualname(Builtin parent, string qualname) {
+  not qualname = "__objclass__" and
+  not qualname.matches("%.%") and
+  result = parent.getMember(qualname)
+  or
+  qualname.matches("%.%") and
+  exists(string before_dot, string after_dot, Builtin intermediate_parent |
+    qualname = before_dot + "." + after_dot and
+    not before_dot = "__objclass__" and
+    intermediate_parent = parent.getMember(before_dot) and
+    result = traverse_qualname(intermediate_parent, after_dot)
+  )
+}
+
+/**
+ * Class of recorded calls where we can identify both the `call` and the `callee` uniquely.
+ */
+class IdentifiedRecordedCall extends XMLRecordedCall {
+  IdentifiedRecordedCall() {
+    strictcount(this.getACall()) = 1 and
+    (
+      strictcount(this.getAPythonCallee()) = 1
+      or
+      strictcount(this.getABuiltinCallee()) = 1
+    )
+    or
+    // Handle case where the same function is called multiple times in one line, for
+    // example `func(); func()`. This only works if:
```

Comment:
Is this actually a common occurrence? (I guess `foo(bar(),bar())` might be, but that doesn't quite fit your definition, from what I understand.)

Reply:
The `bar` part of `foo(bar(), bar())` should also be handled, as long as `bar()` calls the same function.